cephfs-shell: add snapshot management

mchangir opened the pull request:
Snapshots can be managed by:
$ snap {create|delete} snap_name dir_name

Fixes: http://tracker.ceph.com/issues/38681
Signed-off-by: Milind Changire <mchangir@redhat.com>



<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

batrick commented:
LGTM! Now we just need a test for this in `qa/tasks/cephfs/test_cephfs_shell.py`.

mchangir commented:
retest this please

ifleg commented:
instead of using a hardcoded '.snap' you have to get the actual value in the config because it can be modified in ceph.conf... 
thus I suggests:
snapdir = (cephfs.conf_get('client_snapdir') if cephfs.conf_get('client_snapdir') != None else '.snap')
and then use snapdir.

mchangir commented:
> instead of using a hardcoded '.snap' you have to get the actual value in the config because it can be modified in ceph.conf...
> thus I suggests:
> snapdir = (cephfs.conf_get('client_snapdir') if cephfs.conf_get('client_snapdir') != None else '.snap')
> and then use snapdir.

incorporated the configurable snapdir changes

mchangir commented:
> still needs a test @mchangir !

`test_snap()` function has now been added to `qa/tasks/cephfs/test_cephfs_shell.py`

mchangir commented:
> Needs rebased now that #28663 is merged.

done. I've added the test as well.

mchangir commented:
> needs rebased; sorry for the frequent rounds of this. I'll test this in the next batch!

rebased branch

batrick commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1215,6 +1223,42 @@ sub-directories, files')
+        """
+        Snapshot management for the volume
+        """
+        snapdir = (cephfs.conf_get('client_snapdir') if cephfs.conf_get('client_snapdir') is not None else '.snap')
```
Build failure:

````
  6/180 Test  #10: run-tox-cephfs-shell ....................***Failed    6.05 sec
flake8 create: /home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/cephfs/.tox/flake8
flake8 installdeps: flake8
flake8 installed: entrypoints==0.3,flake8==3.7.8,mccabe==0.6.1,pkg-resources==0.0.0,pycodestyle==2.5.0,pyflakes==2.1.1
flake8 runtests: PYTHONHASHSEED='3816273341'
flake8 runtests: commands[0] | flake8 --ignore=W503 --max-line-length=100 cephfs-shell
cephfs-shell:1240:101: E501 line too long (115 > 100 characters)
ERROR: InvocationError: '/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/cephfs/.tox/flake8/bin/flake8 --ignore=W503 --max-line-length=100 cephfs-shell'
````

mchangir replied:
Fixed: `cephfs-shell:1240:101: E501 line too long (115 > 100 characters)`

batrick commented:
please rebase

batrick commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -157,8 +157,13 @@ def get_all_possible_paths(pattern):
-            os.path.join(cephfs.getcwd(), complete_pattern))]
+    pat = ""
+    if is_rel_path:
+        parent_dir = cephfs.getcwd().decode('utf-8')
```
Paths are now always treated as a bytes object. We never decode them unless printing to some external message.

batrick commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1227,6 +1235,45 @@ sub-directories, files')
+                             choices=['create', 'delete'], nargs='?',
+                             default=None)
+    snap_parser.add_argument('name', type=str, help='Name of snapshot')
+    snap_parser.add_argument('dir', type=str,
```
`type=str, action=path_to_bytes`

See other commands and commit history.

varshar16 commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1227,6 +1235,45 @@ sub-directories, files')
+                if is_dir_exists(args.dir):
+                    cephfs.mkdir(os.path.join(args.dir, snapdir, args.name), 0o755)
+                else:
+                    self.poutput("error: no such directory '%s'" % args.dir)
```
For printing errors, please use perror() instead of poutput().

mchangir replied:
> For printing errors, please use perror() instead of poutput().

done.
@varshar16 @batrick please review the changes

varshar16 commented:
Document this feature in doc/cephfs/cephfs-shell.rst

mchangir commented:
> Please fix this error:
> 
> ```
> CephFS:~/>>> mkdir testing
> CephFS:~/>>> snap create test testing
> Can't mix strings and bytes in path components
> Traceback (most recent call last):
>   File "../src/tools/cephfs/cephfs-shell", line 395, in onecmd
>     res = Cmd.onecmd(self, line, **kwargs)
>   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 2148, in onecmd
>     stop = func(statement)
>   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 272, in cmd_wrapper
>     return func(cmd2_app, args)
>   File "../src/tools/cephfs/cephfs-shell", line 1259, in do_snap
>     cephfs.mkdir(os.path.join(args.dir, snapdir, args.name), 0o755)
>   File "/home/varsha/ceph/build/venv/lib64/python3.7/posixpath.py", line 94, in join
>     genericpath._check_arg_types('join', a, *p)
>   File "/home/varsha/ceph/build/venv/lib64/python3.7/genericpath.py", line 151, in _check_arg_types
>     raise TypeError("Can't mix strings and bytes in path components") from None
> TypeError: Can't mix strings and bytes in path components
> CephFS:~/>>> ls
> testing/   
> ```

@batrick @varshar16 
fixed this as well
the snapshot "name" was probably in "string" format that needed to be converted to "bytes"
added "action=path_to_bytes" to the "name" argument as well

mchangir commented:
> Document this feature in doc/cephfs/cephfs-shell.rst

added synopsis of command at bottom of the file

varshar16 commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -157,8 +157,13 @@ def get_all_possible_paths(pattern):
             paths.extend(glob(path, pattern))
-    return [path for path in paths if fnmatch.fnmatch(path,
-            os.path.join(cephfs.getcwd(), complete_pattern))]
+    pat = ""
```
Instead of introducing new variables, we can probably do this:
```suggestion
        if is_rel_path:
        complete_pattern = os.path.join(cephfs.getcwd(), complete_pattern)
    return [path for path in paths if fnmatch.fnmatch(path, complete_pattern)]
```

mchangir replied:
done

varshar16 commented:
> > Please fix this error:
> > ```
> > CephFS:~/>>> mkdir testing
> > CephFS:~/>>> snap create test testing
> > Can't mix strings and bytes in path components
> > Traceback (most recent call last):
> >   File "../src/tools/cephfs/cephfs-shell", line 395, in onecmd
> >     res = Cmd.onecmd(self, line, **kwargs)
> >   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 2148, in onecmd
> >     stop = func(statement)
> >   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 272, in cmd_wrapper
> >     return func(cmd2_app, args)
> >   File "../src/tools/cephfs/cephfs-shell", line 1259, in do_snap
> >     cephfs.mkdir(os.path.join(args.dir, snapdir, args.name), 0o755)
> >   File "/home/varsha/ceph/build/venv/lib64/python3.7/posixpath.py", line 94, in join
> >     genericpath._check_arg_types('join', a, *p)
> >   File "/home/varsha/ceph/build/venv/lib64/python3.7/genericpath.py", line 151, in _check_arg_types
> >     raise TypeError("Can't mix strings and bytes in path components") from None
> > TypeError: Can't mix strings and bytes in path components
> > CephFS:~/>>> ls
> > testing/   
> > ```
> 
> @batrick @varshar16
> fixed this as well
> the snapshot "name" was probably in "string" format that needed to be converted to "bytes"
> added "action=path_to_bytes" to the "name" argument as well

I am still getting the following error:
```
CephFS:~/>>> snap create test dir1
Can't mix strings and bytes in path components
Traceback (most recent call last):
  File "../src/tools/cephfs/cephfs-shell", line 395, in onecmd
    res = Cmd.onecmd(self, line, **kwargs)
  File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 2148, in onecmd
    stop = func(statement)
  File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 272, in cmd_wrapper
    return func(cmd2_app, args)
  File "../src/tools/cephfs/cephfs-shell", line 1260, in do_snap
    cephfs.mkdir(os.path.join(args.dir, snapdir, args.name), 0o755)
  File "/home/varsha/ceph/build/venv/lib64/python3.7/posixpath.py", line 94, in join
    genericpath._check_arg_types('join', a, *p)
  File "/home/varsha/ceph/build/venv/lib64/python3.7/genericpath.py", line 151, in _check_arg_types
    raise TypeError("Can't mix strings and bytes in path components") from None
TypeError: Can't mix strings and bytes in path components
```
```snapdir``` also needs to be converted in bytes.

mchangir commented:
> > > Please fix this error:
> > > ```
> > > CephFS:~/>>> mkdir testing
> > > CephFS:~/>>> snap create test testing
> > > Can't mix strings and bytes in path components
> > > Traceback (most recent call last):
> > >   File "../src/tools/cephfs/cephfs-shell", line 395, in onecmd
> > >     res = Cmd.onecmd(self, line, **kwargs)
> > >   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 2148, in onecmd
> > >     stop = func(statement)
> > >   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 272, in cmd_wrapper
> > >     return func(cmd2_app, args)
> > >   File "../src/tools/cephfs/cephfs-shell", line 1259, in do_snap
> > >     cephfs.mkdir(os.path.join(args.dir, snapdir, args.name), 0o755)
> > >   File "/home/varsha/ceph/build/venv/lib64/python3.7/posixpath.py", line 94, in join
> > >     genericpath._check_arg_types('join', a, *p)
> > >   File "/home/varsha/ceph/build/venv/lib64/python3.7/genericpath.py", line 151, in _check_arg_types
> > >     raise TypeError("Can't mix strings and bytes in path components") from None
> > > TypeError: Can't mix strings and bytes in path components
> > > CephFS:~/>>> ls
> > > testing/   
> > > ```
> > 
> > 
> > @batrick @varshar16
> > fixed this as well
> > the snapshot "name" was probably in "string" format that needed to be converted to "bytes"
> > added "action=path_to_bytes" to the "name" argument as well
> 
> I am still getting the following error:
> 
> ```
> CephFS:~/>>> snap create test dir1
> Can't mix strings and bytes in path components
> Traceback (most recent call last):
>   File "../src/tools/cephfs/cephfs-shell", line 395, in onecmd
>     res = Cmd.onecmd(self, line, **kwargs)
>   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 2148, in onecmd
>     stop = func(statement)
>   File "/home/varsha/ceph/build/venv/lib/python3.7/site-packages/cmd2/cmd2.py", line 272, in cmd_wrapper
>     return func(cmd2_app, args)
>   File "../src/tools/cephfs/cephfs-shell", line 1260, in do_snap
>     cephfs.mkdir(os.path.join(args.dir, snapdir, args.name), 0o755)
>   File "/home/varsha/ceph/build/venv/lib64/python3.7/posixpath.py", line 94, in join
>     genericpath._check_arg_types('join', a, *p)
>   File "/home/varsha/ceph/build/venv/lib64/python3.7/genericpath.py", line 151, in _check_arg_types
>     raise TypeError("Can't mix strings and bytes in path components") from None
> TypeError: Can't mix strings and bytes in path components
> ```
> 
> `snapdir` also needs to be converted in bytes.

done

varshar16 commented:
The snap directory should be displayed on ```ls -a```, right?

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -264,6 +264,67 @@ def test_get_to_console(self):
+        o = self._cephfs_shell("ls -l /data_dir")
+        log.info("cephfs-shell output:\n{}".format(o))
+
+        # create the snapshot - must pass
```
It would be nicer if we can move each testcase into different methods.

mchangir replied:
since the test methods run sequence cannot be guaranteed its difficult to split the test into different methods

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -264,6 +264,67 @@ def test_get_to_console(self):
         log.info("o_hash:{}".format(o_hash))
         assert(s_hash == o_hash)
 
+    def test_snap(self):
```
Would be better to not to place `test_snap` under `TestGetAndPut`.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -264,6 +264,67 @@ def test_get_to_console(self):
         log.info("o_hash:{}".format(o_hash))
         assert(s_hash == o_hash)
 
+    def test_snap(self):
```
Would be nicer to not to put `test_snap` under `TestGetAndPut`. Create a new class `TestSnapshots` instead?

mchangir replied:
done ... separate class created

batrick commented:
@mchangir I think you wanted me to look at this again (mentioned in standup?) but there is a test failure you haven't addressed?

mchangir commented:
> ```
> 2019-09-22T11:12:48.278 INFO:tasks.cephfs_test_runner:======================================================================
> 2019-09-22T11:12:48.279 INFO:tasks.cephfs_test_runner:ERROR: test_snap (tasks.cephfs.test_cephfs_shell.TestGetAndPut)
> 2019-09-22T11:12:48.279 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2019-09-22T11:12:48.279 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2019-09-22T11:12:48.280 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20190922.042910/qa/tasks/cephfs/test_cephfs_shell.py", line 300, in test_snap
> 2019-09-22T11:12:48.281 INFO:tasks.cephfs_test_runner:    self._cephfs_shell("mkdir data_dir")
> 2019-09-22T11:12:48.281 INFO:tasks.cephfs_test_runner:AttributeError: 'TestGetAndPut' object has no attribute '_cephfs_shell'
> ```
> 
> From: /ceph/teuthology-archive/pdonnell-2019-09-22_10:05:30-fs-wip-pdonnell-testing-20190922.042910-distro-basic-smithi/4325928/teuthology.log

changed the calls to the new method calls

mchangir commented:
@batrick @varshar16 @rishabh-d-dave Please review the updated PR

mchangir commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1245,6 +1249,50 @@ sub-directories, files')
+        """
+        # setting self.colors to None turns off colorizing and
+        # perror emits plain text
+        self.colors = None
```
`perror()` emits the ANSI escapes for coloring terminal output ... and that gets in the way of parsing text on `stderr`.
Also, there's no other way to quiesce emission of color codes that I could find.

rishabh-d-dave replied:
Yeah, but just turning colouring off looks like an extreme measure. You reset the value at end of method and add a TODO. 

Anyways, I have written a PR that would allow turning colouring. You can also use regex till then (that is what I did as temporary workaround).

batrick replied:
@rishabh-d-dave let's clean this up in your coloring PR.

batrick commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -48,6 +48,11 @@ def run_cephfs_shell_cmd(self, cmd, mount_x=None, opts=None, stdin=None):
+    def get_cephfs_shell_cmd_error(self, cmd, mount_x=None, opts=None,
+                                    stdin=None):
+        return self.run_cephfs_shell_cmd(cmd, mount_x, opts, stdin).stderr.\
+            getvalue().strip()
```
You need to modify `run_cephfs_shell_cmd` to save stderr:

`mount_x.client_remote.run(args=args, stdout=StringIO(),`

to

`mount_x.client_remote.run(args=args, stdout=StringIO(), stderr=StringIO(),`

Otherwise we see this failure:

````
2019-11-13T13:52:20.985 INFO:tasks.cephfs.test_cephfs_shell:Running command: cephfs-shell -c /etc/ceph/ceph.conf -- snap create snap1 /data_dir
2019-11-13T13:52:20.985 INFO:teuthology.orchestra.run.smithi203:Running:
2019-11-13T13:52:20.986 INFO:teuthology.orchestra.run.smithi203:> cephfs-shell -c /etc/ceph/ceph.conf -- 'snap create snap1 /data_dir'
2019-11-13T13:52:21.208 INFO:teuthology.orchestra.run.smithi203.stderr:ERROR: snapshot 'snap1' already exists
2019-11-13T13:52:21.208 INFO:teuthology.orchestra.run.smithi203.stderr:To enable full traceback, run the following command:  'set debug true'
2019-11-13T13:52:21.232 INFO:tasks.cephfs_test_runner:test_snap (tasks.cephfs.test_cephfs_shell.TestSnapshots) ... ERROR
...
2019-11-13T13:52:50.152 INFO:tasks.cephfs_test_runner:======================================================================
2019-11-13T13:52:50.152 INFO:tasks.cephfs_test_runner:ERROR: test_snap (tasks.cephfs.test_cephfs_shell.TestSnapshots)
2019-11-13T13:52:50.152 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-11-13T13:52:50.152 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-11-13T13:52:50.152 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191112.231923/qa/tasks/cephfs/test_cephfs_shell.py", line 334, in test_snap
2019-11-13T13:52:50.152 INFO:tasks.cephfs_test_runner:    o = self.get_cephfs_shell_cmd_error("snap create snap1 /data_dir")
2019-11-13T13:52:50.153 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191112.231923/qa/tasks/cephfs/test_cephfs_shell.py", line 53, in get_cephfs_shell_cmd_error
2019-11-13T13:52:50.153 INFO:tasks.cephfs_test_runner:    return self.run_cephfs_shell_cmd(cmd, mount_x, opts, stdin).stderr.\
2019-11-13T13:52:50.153 INFO:tasks.cephfs_test_runner:AttributeError: 'NoneType' object has no attribute 'getvalue'
````

From: /ceph/teuthology-archive/pdonnell-2019-11-13_04:26:42-fs-wip-pdonnell-testing-20191112.231923-distro-basic-smithi/4502318/teuthology.log

mchangir replied:
@batrick fyi - [4545412](http://qa-proxy.ceph.com/teuthology/mchangir-2019-11-26_12:17:58-fs-wip-cephf-shell-add-snapshot-management-distro-basic-smithi/4545412/teuthology.log), [4545413](http://qa-proxy.ceph.com/teuthology/mchangir-2019-11-26_12:17:58-fs-wip-cephf-shell-add-snapshot-management-distro-basic-smithi/4545413/teuthology.log)

Things look okay here ?

batrick replied:
Odd... resolving this.

batrick commented:
@mchangir please run this through teuthology with `--filter cephfs-shell` when you are done to confirm it works.
